File: cpp-sdk/src/esg.cpp
#include "erc8040/esg.hpp"
#include <algorithm>
#include <cmath>

namespace erc8040 {

namespace {
uint8_t clamp_score(double value) {
    double clamped = std::clamp(value, 0.0, 100.0);
    return static_cast<uint8_t>(std::lround(clamped));
}
}  // namespace
bool ESGScore::is_investment_grade() const {
    return rating == ESGRating::AAA || 
           rating == ESGRating::AA || 
           rating == ESGRating::A || 
           rating == ESGRating::BBB;
}

ESGScoring::ESGScoring(double env_weight, double social_weight, double gov_weight)
    : env_weight_(env_weight), social_weight_(social_weight), gov_weight_(gov_weight) {
    if (env_weight < 0.0 || social_weight < 0.0 || gov_weight < 0.0) {
        throw std::invalid_argument("ESG weights must be non-negative");
    }

    double total = env_weight + social_weight + gov_weight;
    if (total <= 0.0) {
        throw std::invalid_argument("ESG weights must sum to > 0");
    }

    env_weight_ /= total;
    social_weight_ /= total;
    gov_weight_ /= total;
}

ESGScore ESGScoring::calculate(double environmental, double social, double governance) const {
    uint8_t environmental_score = clamp_score(environmental);
    uint8_t social_score = clamp_score(social);
    uint8_t governance_score = clamp_score(governance);

    double weighted = environmental_score * env_weight_ + 
                      social_score * social_weight_ + 
                      governance_score * gov_weight_;
    uint8_t total = clamp_score(weighted);
    ESGRating rating = rating_from_score(total);
    
    return ESGScore{
        environmental_score,
        social_score,
        governance_score,
        total,
        rating
    };
}

ESGRating ESGScoring::rating_from_score(uint8_t score) {
    if (score >= 90) return ESGRating::AAA;
    if (score >= 85) return ESGRating::AA;
    if (score >= 80) return ESGRating::A;
    if (score >= 70) return ESGRating::BBB;
    if (score >= 60) return ESGRating::BB;
    if (score >= 50) return ESGRating::B;
    if (score >= 40) return ESGRating::CCC;
    if (score >= 30) return ESGRating::CC;
    if (score >= 20) return ESGRating::C;
    return ESGRating::D;
}

std::string ESGScoring::rating_to_string(ESGRating rating) {
    switch (rating) {
        case ESGRating::AAA: return "AAA";
        case ESGRating::AA: return "AA";
        case ESGRating::A: return "A";
        case ESGRating::BBB: return "BBB";
        case ESGRating::BB: return "BB";
        case ESGRating::B: return "B";
        case ESGRating::CCC: return "CCC";
        case ESGRating::CC: return "CC";
        case ESGRating::C: return "C";
        case ESGRating::D: return "D";
        default: return "Unknown";
    }
}

} // namespace erc8040

File: python-sdk/src/erc8040_sdk/esg.py
"""ESG scoring and rating module for ERC-8040."""
from enum import Enum

from pydantic import BaseModel, Field


class ESGCategory(str, Enum):
    """ESG Category types."""

    ENVIRONMENTAL = "ENVIRONMENTAL"
    SOCIAL = "SOCIAL"
    GOVERNANCE = "GOVERNANCE"


class ESGRating(str, Enum):
    """ESG Rating levels from D (lowest) to AAA (highest)."""

    D = "D"
    C = "C"
    CC = "CC"
    CCC = "CCC"
    B = "B"
    BB = "BB"
    BBB = "BBB"
    A = "A"
    AA = "AA"
    AAA = "AAA"

    @classmethod
    def from_score(cls, score: float) -> "ESGRating":
        """Convert a total ESG score (0-100) to a rating."""
        if score >= 90.0:
            return cls.AAA
        elif score >= 85.0:
            return cls.AA
        elif score >= 80.0:
            return cls.A
        elif score >= 70.0:
            return cls.BBB
        elif score >= 60.0:
            return cls.BB
        elif score >= 50.0:
            return cls.B
        elif score >= 40.0:
            return cls.CCC
        elif score >= 30.0:
            return cls.CC
        elif score >= 20.0:
            return cls.C
        else:
            return cls.D

    @property
    def is_investment_grade(self) -> bool:
        """Check if this rating is investment grade (BBB or higher)."""
        return self in (ESGRating.AAA, ESGRating.AA, ESGRating.A, ESGRating.BBB)


class ESGScore(BaseModel):
    """ESG Score breakdown."""

    environmental: float = Field(ge=0.0, le=100.0)
    social: float = Field(ge=0.0, le=100.0)
    governance: float = Field(ge=0.0, le=100.0)
    total: float = Field(ge=0.0, le=100.0)
    rating: ESGRating

    @classmethod
    def create(cls, environmental: float, social: float, governance: float) -> "ESGScore":
        """Create a new ESG score with calculated total and rating."""
        total = (environmental + social + governance) / 3.0
        rating = ESGRating.from_score(total)
        return cls(
            environmental=environmental,
            social=social,
            governance=governance,
            total=total,
            rating=rating,
        )

    def is_investment_grade(self) -> bool:
        """Check if this score is investment grade."""
        return self.rating.is_investment_grade


class ESGScoring:
    """ESG Scoring calculator with configurable weights."""

    def __init__(
        self,
        environmental_weight: float = 1.0 / 3.0,
        social_weight: float = 1.0 / 3.0,
        governance_weight: float = 1.0 / 3.0,
    ):
        """Initialize ESG scoring with custom weights.

        Args:
            environmental_weight: Weight for environmental score (default: 1/3)
            social_weight: Weight for social score (default: 1/3)
            governance_weight: Weight for governance score (default: 1/3)
        """
        total = environmental_weight + social_weight + governance_weight
        self.environmental_weight = environmental_weight / total
        self.social_weight = social_weight / total
        self.governance_weight = governance_weight / total

    def calculate(self, environmental: float, social: float, governance: float) -> ESGScore:
        """Calculate ESG score from individual component scores.

        Args:
            environmental: Environmental score (0-100)
            social: Social score (0-100)
            governance: Governance score (0-100)

        Returns:
            ESGScore with weighted total and rating
        """
        weighted_total = (
            environmental * self.environmental_weight
            + social * self.social_weight
            + governance * self.governance_weight
        )

        rating = ESGRating.from_score(weighted_total)

        return ESGScore(
            environmental=environmental,
            social=social,
            governance=governance,
            total=weighted_total,
            rating=rating,
        )
